dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/RouterChain.java

```diff
@@ -111,6 +203,96 @@ private void sort() {
+        stateRouters.forEach(router -> router.notify(this.invokers));
         routers.forEach(router -> router.notify(this.invokers));
+        loop(true);
+    }
+
+    private void buildCache(boolean notify) {
+        if (invokers == null || invokers.size() <= 0) {
+            return;
+        }
+        AddrCache origin = cache.get();
+        List copyInvokers = new ArrayList<>(this.invokers);
+        CountDownLatch cdl = new CountDownLatch(stateRouters.size());
+        AddrCache newCache = new AddrCache();
+        newCache.setInvokers((List)invokers);
+        final AtomicBoolean poolSuccess = new AtomicBoolean(true);
+        for (StateRouter stateRouter : (List<StateRouter>)stateRouters) {
+            if (stateRouter.isEnable()) {
+                executorRepository.getPoolRouterExecutor().execute(new Runnable() {
+                    @Override
+                    public void run() {
+                        RouterCache routerCache = null;
+                        try {
+                            routerCache = poolRouter(stateRouter, origin, copyInvokers, notify);
+                            //file cache
+                            newCache.getCache().put(stateRouter.getName(), routerCache);
+                        } catch (Throwable t) {
+                            poolSuccess.set(false);
+                            logger.error("Failed to pool router: " + stateRouter.getUrl() + ", cause: " + t.getMessage(), t);
+                        } finally {
+                            cdl.countDown();
+                        }
+                    }
+                });
+            }
+        }
+        try {
+            cdl.await();
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+        }
+
+        if (poolSuccess.get()) {
+            this.cache.set(newCache);
+        }
+    }
+
+    private RouterCache poolRouter(StateRouter router, AddrCache orign, List<Invoker<T>> invokers, boolean notify) {
+        String routerName = router.getName();
+        RouterCache routerCache = null;
+        if (isCacheMiss(orign, routerName) || router.shouldRePool() || notify) {
+            return router.pool(invokers);
+        } else {
+            routerCache = orign.getCache().get(routerName);
+        }
+        if (routerCache == null) {
+            return new RouterCache();
+        }
+        return routerCache;
+    }
+
+    private boolean isCacheMiss(AddrCache cache, String routerName) {
+        if (cache == null || cache.getCache() == null || cache.getInvokers() == null || cache.getCache().get(routerName) == null) {
+            return true;
+        }
+        return false;
+    }
+
+    public void loop(boolean notify) {
+        // 1、多个服务并行执行 buildCache
+        // 2、notify true/false
+        if (notify) {
+            if (loopPermitNotify.tryAcquire()) {
+                LOOP_THREAD_POOL.submit(new Runnable() {
```

Comment:
Try make this Runnable as a static class reference or lambda reference. Reduce inner class creation.